### PR TITLE
Login flow

### PR DESCRIPTION
**Pull request by vvaditya23**
https://github.com/user-attachments/assets/7b188522-7427-44df-a608-9f037d85c89e

### CODE REVIEW

File: iOSProjectShaadi.xcodeproj/project.xcworkspace/contents.xcworkspacedata
Comment (Pulkit29) [inline iOSProjectShaadi.xcodeproj/project.xcworkspace/contents.xcworkspacedata:1]: We should not push this file. Add this to .gitignore

File: iOSProjectShaadi.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
Comment (Pulkit29) [inline iOSProjectShaadi.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: Add this to gitignore

File: iOSProjectShaadi/Assets.xcassets/Ellipse 39.imageset/Contents.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "Ellipse 39.png",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 39.imageset/Contents.json]: we should use 2x/3x not 1x or use single scale with pdf assets

File: iOSProjectShaadi/Assets.xcassets/Ellipse 46.imageset/Contents.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 46.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/Ellipse 47.imageset/Contents.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "Ellipse 47.png",
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 47.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/Ellipse 48.imageset/Contents.json [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "Ellipse 48.png",
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 48.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/Ellipse 49.imageset/Contents.json [hunk 5]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "Ellipse 49.png",
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 49.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/Ellipse 53.imageset/Contents.json [hunk 6]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "Ellipse 53.png",
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 53.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/Ellipse 58.imageset/Contents.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "Ellipse 58.png",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 58.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/Ellipse 63.imageset/Contents.json [hunk 8]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "Ellipse 63.png",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/Ellipse 63.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/apple.logo.imageset/Contents.json [hunk 9]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "apple.logo.png",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/apple.logo.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/arrow.right.imageset/Contents.json [hunk 10]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "arrow.right.png",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/arrow.right.imageset/Contents.json]: same

File: iOSProjectShaadi/Assets.xcassets/google.logo.imageset/Contents.json [hunk 11]
```diff
@@ -0,0 +1,21 @@
+{
+  "images" : [
+    {
+      "filename" : "google.logo.png",
+      "idiom" : "universal",
+      "scale" : "1x"
+    },
+    {
+      "idiom" : "universal",
```
Comment (Pulkit29) [inline iOSProjectShaadi/Assets.xcassets/google.logo.imageset/Contents.json]: same

File: iOSProjectShaadi/Onboarding/Login/LoginViewController.swift [hunk 12]
```diff
@@ -0,0 +1,101 @@
+//
+//  LoginViewController.swift
+//  iOSProjectShaadi
+//
+//  Created by Aditya Vyavahare on 19/07/24.
+//
+
+import UIKit
+
+class LoginViewController: UIViewController {
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/LoginViewController.swift]: final

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+            textField.heightAnchor.constraint(equalToConstant: 60).isActive = true
+            textField.widthAnchor.constraint(equalToConstant: 60).isActive = true
+        }
+        
+        otpStackView.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(otpStackView)
+        
+        NSLayoutConstraint.activate([
+            titleLabel.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 124),
+            titleLabel.leftAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leftAnchor, constant: 63),
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: Contraints are not complete. We need to complete all 4 sides.

File: iOSProjectShaadi/Onboarding/Login/LoginViewController.swift [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    // UI elements
+    let titleLabel = UILabel()
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/LoginViewController.swift]: Make all UI elements as  computed property and set relevant attributes here itself during computation.

File: iOSProjectShaadi/Onboarding/Login/LoginViewController.swift [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+        otpButton.translatesAutoresizingMaskIntoConstraints = false
+        passwordButton.translatesAutoresizingMaskIntoConstraints = false
+
+        view.addSubview(titleLabel)
+        view.addSubview(subtitleLabel)
+        view.addSubview(numberTextField)
+        view.addSubview(otpButton)
+        view.addSubview(passwordButton)
+
+        NSLayoutConstraint.activate([
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/LoginViewController.swift:88]: Contraints should be set from viewDidLayoutSubviews method

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift [hunk 16]
```diff
@@ -0,0 +1,145 @@
+//
+//  OTPViewController.swift
+//  iOSProjectShaadi
+//
+//  Created by Aditya Vyavahare on 29/07/24.
+//
+
+import UIKit
+
+class OTPViewController: UIViewController {
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: final

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+    let titleLabel = UILabel()
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: same as in Login VC

File: iOSProjectShaadi/Onboarding/Login/LoginViewController.swift (hunk 15 above)
Comment (Pulkit29) [reply iOSProjectShaadi/Onboarding/Login/LoginViewController.swift:88]: Constraints are not complete.

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        view.addSubview(submitButton)
+        
+        let otpStackView = UIStackView()
+        otpStackView.axis = .horizontal
+        otpStackView.alignment = .fill
+        otpStackView.distribution = .fillEqually
+        otpStackView.spacing = 10
+        
+        for textFieldNumber in 0..<4 {
+            let textField = UITextField()
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: We should create a separate customView for this

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+// MARK: Delegate conformance
+extension OTPViewController: UITextFieldDelegate {
+    func textFieldDidBeginEditing(_ textField: UITextField) {
+        textField.layer.borderColor = UIColor(red: 221/255, green: 87/255, blue: 82/255, alpha: 1).cgColor
+    }
+    func textFieldDidEndEditing(_ textField: UITextField) {
+        textField.layer.borderColor = UIColor.label.cgColor
+    }
+    
+    @objc func textFieldDidChange(_ textField: UITextField) {
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: private

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+        
+        submitButton.setTitle("Next", for: .normal)
+        submitButton.setTitleColor(.white, for: .normal)
+        submitButton.titleLabel?.font = UIFont(name: "KumbhSans-SemiBold", size: 16)
+        submitButton.layer.cornerRadius = 20
+        submitButton.backgroundColor = UIColor(red: 221/255, green: 87/255, blue: 82/255, alpha: 1)
+    }
+}
+
+// MARK: Delegate conformance
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: UITextFieldDelegate conformance

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeItem.swift [hunk 21]
```diff
@@ -0,0 +1,13 @@
...
+//  WelcomeItem.swift
+//  iOSProjectShaadi
+//
+//  Created by Rathi, Radhika on 24/07/2024.
+//
+
+import Foundation
+
+struct WelcomeItem {
+//  var subTitle: String
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeItem.swift:11]: remove commented code

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeView.swift [hunk 22]
```diff
@@ -0,0 +1,302 @@
+//
+//  WelcomeView.swift
+//  iOSProjectShaadi
+//
+//  Created by Rathi, Radhika on 24/07/2024.
+//
+
+import UIKit
+
+class WelcomeView: UIView {
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeView.swift:10]: final

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeView.swift [hunk 23]
```diff
@@ -0,0 +1,302 @@
...
+      appleSignInButton.topAnchor.constraint(equalTo: getStartedButton.topAnchor),
+      appleSignInButton.leadingAnchor.constraint(equalTo: googleSignInButton.trailingAnchor, constant: 10),
+      appleSignInButton.widthAnchor.constraint(equalToConstant: 57),
+      appleSignInButton.heightAnchor.constraint(equalToConstant: 60),
+
+      accountLabel.topAnchor.constraint(equalTo: getStartedButton.bottomAnchor, constant: 20),
+      accountLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 20),
+      accountLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -20),
+
+      loginButton.topAnchor.constraint(equalTo: getStartedButton.bottomAnchor, constant: 20),
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeView.swift:251]: please attach screenshot for this screen

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeViewController.swift [hunk 24]
```diff
@@ -0,0 +1,73 @@
+//
+//  WelcomeViewController.swift
+//  iOSProjectShaadi
+//
+//  Created by Rathi, Radhika on 24/07/2024.
+//
+
+import UIKit
+
+class WelcomeViewController: UIViewController {
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeViewController.swift:10]: final

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeViewController.swift [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+  private let welcomeView = WelcomeView()
+  private let viewModel: WelcomeViewModel
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeViewController.swift:12]: create a protocol for WelcomeViewModel and WelcomeView

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeViewModel.swift [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+//  WelcomeViewModel.swift
+//  iOSProjectShaadi
+//
+//  Created by Rathi, Radhika on 31/07/2024.
+//
+
+import Foundation
+import GoogleSignIn
+
+class WelcomeViewModel {
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeViewModel.swift:11]: final

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeViewModel.swift [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+    self.items = items
+  }
+
+  func handleGetStarted() {
+
+    navigateToSignUp?()
+    print("Get Started button action triggered in ViewModel")
+  }
+
+  func handleGoogleSignIn(presenter: UIViewController) {
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeViewModel.swift:33]: create a separate handler or socialMediaAuthenticationHandler and inject in ViewModel. Keep separation of concern

File: iOSProjectShaadi/SceneDelegate.swift [hunk 28]
```diff
@@ -0,0 +1,68 @@
...
+
+        WelcomeItem(description: "Join us and find your perfect match among millions."),
+      ]
+
+      window = UIWindow(windowScene: windowScene)
+
+      let viewModel = WelcomeViewModel(items: items)
+
+      let welcomeViewController = WelcomeViewController(viewModel: viewModel)
+      let navigationController = UINavigationController(rootViewController: welcomeViewController)
```
Comment (Pulkit29) [inline iOSProjectShaadi/SceneDelegate.swift:30]: Try to use Coordinator pattern. Create one AppRouter and use here

File: iOSProjectShaadi/Onboarding/Welcome/WelcomeView.swift [hunk 29]
```diff
@@ -0,0 +1,302 @@
...
+  }
+
+  func configure(with item: WelcomeItem) {
+//    subtitleLabel.text = item.subTitle
+    descriptionLabel.text = item.description
+  }
+
+  var getStartedButtonAction: (() -> Void)? {
+    didSet {
+      getStartedButton.addTarget(self, action: #selector(getStartedButtonTapped), for: .touchUpInside)
```
Comment (Pulkit29) [inline iOSProjectShaadi/Onboarding/Welcome/WelcomeView.swift:264]: Set target at the time of button declaration itself

File: iOSProjectShaadi.xcodeproj/project.xcworkspace/contents.xcworkspacedata
Comment (vvaditya23) [reply iOSProjectShaadi.xcodeproj/project.xcworkspace/contents.xcworkspacedata:1]: `*.xcworkspace` added to .gitignore file

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift (hunk 13 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: Done

File: iOSProjectShaadi/Onboarding/Login/LoginViewController.swift (hunk 15 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/LoginViewController.swift:88]: Done

File: iOSProjectShaadi.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist
Comment (vvaditya23) [reply iOSProjectShaadi.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist:1]: `*.xcworkspace` added to .gitignore file

File: iOSProjectShaadi/Onboarding/Login/LoginViewController.swift (hunk 12 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/LoginViewController.swift]: Done

File: iOSProjectShaadi/Onboarding/Login/LoginViewController.swift (hunk 14 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/LoginViewController.swift]: Done

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift (hunk 16 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: Done

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift (hunk 17 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: Done

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift (hunk 18 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: Implemented custom classes `OTPTextField` & `OTPStackView` 
This file can be found here `/iOSProjectShaadi/Utilities/OTPTextField.swift`

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift (hunk 19 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: This code block is no more required hence removed it

File: iOSProjectShaadi/Onboarding/Login/OTPViewController.swift (hunk 20 above)
Comment (vvaditya23) [reply iOSProjectShaadi/Onboarding/Login/OTPViewController.swift]: Removed this

**Pull request closed without merging**